src/elements/element.rectangle.js

```diff
@@ -78,70 +83,92 @@ module.exports = Element.extend({
-		}
-
-		// Canvas doesn't allow us to stroke inside the width so we can
-		// adjust the sizes to fit if we're setting a stroke on the line
-		if (borderWidth) {
-			// borderWidth shold be less than bar width and bar height.
-			var barSize = Math.min(Math.abs(left - right), Math.abs(top - bottom));
-			borderWidth = borderWidth > barSize ? barSize : borderWidth;
-			var halfStroke = borderWidth / 2;
-			// Adjust borderWidth when bar top position is near vm.base(zero).
-			var borderLeft = left + (borderSkipped !== 'left' ? halfStroke * signX : 0);
-			var borderRight = right + (borderSkipped !== 'right' ? -halfStroke * signX : 0);
-			var borderTop = top + (borderSkipped !== 'top' ? halfStroke * signY : 0);
-			var borderBottom = bottom + (borderSkipped !== 'bottom' ? -halfStroke * signY : 0);
-			// not become a vertical line?
-			if (borderLeft !== borderRight) {
-				top = borderTop;
-				bottom = borderBottom;
-			}
-			// not become a horizontal line?
-			if (borderTop !== borderBottom) {
-				left = borderLeft;
-				right = borderRight;
-			}
+			borderSkipped = valueOrDefault(vm.borderSkipped, 'left') || '';
 		}
 
-		ctx.beginPath();
 		ctx.fillStyle = vm.backgroundColor;
 		ctx.strokeStyle = vm.borderColor;
-		ctx.lineWidth = borderWidth;
-
-		// Corner points, from bottom-left to bottom-right clockwise
-		// | 1 2 |
-		// | 0 3 |
-		var corners = [
-			[left, bottom],
-			[left, top],
-			[right, top],
-			[right, bottom]
-		];
-
-		// Find first (starting) corner with fallback to 'bottom'
-		var borders = ['bottom', 'left', 'top', 'right'];
-		var startCorner = borders.indexOf(borderSkipped, 0);
-		if (startCorner === -1) {
-			startCorner = 0;
+
+		if (!borderWidth) {
+			ctx.fillRect(left, bottom, right - left, top - bottom);
+			return;
 		}
 
-		function cornerAt(index) {
-			return corners[(startCorner + index) % 4];
+		maxWidth = Math.abs(left - right) / 2;
+		maxHeight = Math.abs(top - bottom) / 2;
+		if (helpers.isObject(borderWidth)) {
+			borderWidth = {
+				bottom: Math.min(borderWidth.bottom || 0, maxHeight),
+				left: Math.min(borderWidth.left || 0, maxWidth),
+				top: Math.min(borderWidth.top || 0, maxHeight),
+				right: Math.min(borderWidth.right || 0, maxWidth)
+			};
+		} else {
+			borderWidth = {
+				bottom: borderSkipped === 'bottom' ? 0 : Math.min(borderWidth, maxHeight),
+				left: borderSkipped === 'left' ? 0 : Math.min(borderWidth, maxWidth),
+				top: borderSkipped === 'top' ? 0 : Math.min(borderWidth, maxHeight),
+				right: borderSkipped === 'right' ? 0 : Math.min(borderWidth, maxWidth)
+			};
 		}
 
-		// Draw rectangle from 'startCorner'
-		var corner = cornerAt(0);
-		ctx.moveTo(corner[0], corner[1]);
+		ctx.fillRect(
+			left + borderWidth.left * signX,
+			bottom - borderWidth.bottom * signY,
+			right - left - signX * (borderWidth.left + borderWidth.right),
+			top - bottom + signY * (borderWidth.top + borderWidth.bottom));
+
+		function drawBorder(w, x1, y1, x2, y2) {
```

Comment:
Most places in the codebase we declare functions at the top of the file. I think the only place I've seen it done within another function is when the helper function is only a few lines long. Do we have one style or the other that we should prefer?

Reply:
This is a internal helper, relying on local variables. If it would be used in another function, refactoring outside would be good idea.